### PR TITLE
fix(runner): store the SqliteDb handle self-contained (multi-runtime shared-query hash stability)

### PR DESCRIPTION
**Pull request by seefeldb**
## Problem

A shared (space-scoped) sqlite query result cell ping-pongs indefinitely between two runtimes when the db's rowLabel rule uses a term LIST (e.g. `all(a, b, c)`).

Mechanism: the SqliteDb handle value was not stored **self-contained**. `sqliteDatabase`'s first action run can execute inside the pattern's builder frame, where `Cell.set` [ID]-anchors every object-in-array — splitting a rule's term list into per-element entity docs — and a proxy-carried `tables` read stores as a LINK into the pattern's doc graph. No schema-driven sync ever loads those linked docs, so a second runtime deep-resolves them to `null` (the #3830 class) and `sqliteQuery` hashes `rowLabel.confidentiality.allOf: [null]` while the creator hashes the resolved AST. Two request hashes on ONE shared result cell → each runtime sees the other's hash as "new inputs" and re-issues forever; the query never settles (~400 action re-runs in seconds) and nothing ever loads the term docs in the second runtime.

This affects any real multi-client use of a rule-bearing db whose rule has 2+ terms — e.g. `packages/patterns/cfc-row-label-mailbox` (`all(sender, recipients, dbOwner())`).

## Fix

Make the stored handle **self-contained**, and its re-derivation inert:

- **`sqliteDatabase` writes the handle RAW and inline** (`setRawUntyped(fabricFromNativeValue(...), onlyIfDifferent)`) after materializing the merged `tables` to plain JSON. No [ID] splits, no proxy→link capture: every runtime that can read the handle doc has the full resolved spec, so request hashes agree by construction.
- **Fail closed on an unresolved materialization**: a loader runtime that deep-reads `null` through a not-yet-loaded linked doc must not overwrite a good prior value with a null-riddled rule. `validateRowLabelSpec` doubles as the resolution probe (a resolved rule always validates; a null-riddled one never does) — the prior handle value stands.
- **The re-derivation carries `db.exec`'s `rev` forward** — it used to delete it, which every handle hasher (`reactOn: db`) saw as "new inputs": one spurious re-issue of the shared query per runtime init.
- **`db.exec` bumps `rev` as a LEAF write** instead of rewriting the whole handle value — the whole-value `.set()` ran inside the handler frame and re-split the inline term lists on the first write.

Old handles (split/link form) self-heal: the next runtime that fully resolves them rewrites the inline form; `readDbRef` / `db.exec` keep their materializing reads for the transition.

## Red-green

`packages/runner/test/sqlite-handle-multi-runtime.test.ts` — two real Runtimes with **separate storage managers over one in-process memory server** (separate heaps; the single-manager `emulate()` setup shares the heap and masks the bug).

- *"stores the handle value self-contained"*: pre-fix the term list stores as linked docs (`allOf: [<link>, <link>]`), and `db.exec` re-splits it; post-fix the raw handle contains no links, before and after an exec.
- *"a second runtime adopts the settled shared query instead of re-issuing"*: pre-fix a hydrating second runtime rewrites the handle and re-issues the settled query under a new hash (invalidating the first runtime's hash); post-fix it issues **zero** `sqliteQuery` RPCs and both runtimes hold one unchanged request hash.

Both steps verified red against unfixed source (stash/run/pop) and green with the fix. Full `packages/runner` suite: 748 passed, 0 failed.

## Notes

- Spec updated: `docs/specs/sqlite-builtin/03-database-sources.md` now records the self-containment contract.
- Follow-up for #4484: once this lands, the `sqlite-read-clearance` fixture can flip back from its bare confidentiality term to the natural term-list rule (its comment tracks this issue).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Fixes a multi-runtime ping-pong where shared `sqliteQuery` results kept re-issuing due to linked rule term lists. The `SqliteDb` handle is now stored self-contained, so request hashes stay stable across runtimes.

- Bug Fixes
  - Store the handle inline via a raw write after materializing `tables` to plain JSON, preventing link capture and [ID]-splits of rowLabel term lists; all runtimes hash the same inputs.
  - Fail closed on unresolved materialization to avoid overwriting a good handle with `null`-riddled rules; use `validateRowLabelSpec` to detect resolution before writing.
  - Keep `rev` stable: carry it forward during re-derivation and bump it via a leaf write in `db.exec` to avoid hash churn and re-splitting. Also added a multi-runtime test to confirm zero new `sqliteQuery` RPCs and updated the spec to document the self-containment contract.

<sup>Written for commit ae2eca963a5f1af0b09b8a073825a0ad3e2e4c78. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/commontoolsinc/labs/pull/4509?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

- Overlaps textually with #4506 (owner re-mint fix) in the same `sqliteDatabase` hunk; the two compose semantically. Whichever lands second resolves the conflict by keeping #4506's `owner = prior !== undefined ? prior.owner : actingPrincipal` derivation AND this PR's raw inline write (the derived `owner` just feeds the written object).

### CODE REVIEW

File: packages/runner/src/builtins/sqlite-builtins.ts [hunk 1]
```diff
@@ -493,13 +531,44 @@ export function sqliteDatabase(
...
+        // an unchanged re-derivation write-free (no hash churn per runtime).
+        handle.withTx(tx).setRawUntyped(
+          fabricFromNativeValue({
+            id,
+            ...(tables !== undefined && { tables }),
+            scope,
+            ...(owner !== undefined && { owner }),
+            // Carry db.exec's write revision forward — dropping it would make
+            // this re-derivation look like "new inputs" to every handle hasher.
+            ...(typeof prior?.rev === "number" && { rev: prior.rev }),
```
[the remark this answers is not in the document]
Comment (seefeldb) [reply packages/runner/src/builtins/sqlite-builtins.ts:567]: Not valid, I believe: `readDbRef` dropped `rev` from the hashed db reference before this PR too — this change doesn't alter what the hash covers. Write-triggered refresh is (by design) opt-in via `reactOn: db`, which hashes the handle's raw deep value *including* `rev`, so an exec still re-issues such queries — verified empirically in this PR's development (the request hash changes across a seed write). A query without `reactOn` was never auto-refreshed on writes (see the note in sqlite-builtins.test.ts: read-after-write reactivity is intentionally not auto-driven), and folding `rev` into `readDbRef` would silently flip that design for every query. The leaf-rev write also keeps the pre-existing dependency behavior: the query action still re-runs on the handle-doc change and dedups on an unchanged hash.
